src/pybind/mgr/ssh/module.py

```diff
@@ -1393,3 +1395,84 @@ def _remove_rbd_mirror(daemons):
 
     def update_rbd_mirror(self, spec):
         return self._update_service('rbd-mirror', self.add_rbd_mirror, spec)
+
+    def _get_container_image_id(self, image_name):
+        # pick a random host...
+        host = None
+        for host_name in self.inventory_cache:
+            host = host_name
+            break
+        if not host:
+            raise OrchestratorError('no hosts defined')
+        self.log.debug('using host %s' % host)
+        out, code = self._run_ceph_daemon(
+            host, None, 'pull', [],
+            image=image_name,
+            no_fsid=True)
+        return out[0]
+
+    def upgrade_check(self, target):
+        return self._get_services().then(lambda daemons: self._upgrade_check(target, daemons))
+
+    def _upgrade_check(self, target, services):
+        # get service state
+        r = {
+            'global_target_image': target,
+            'needs_update': dict(),
+            'up_to_date': list(),
+        }
+
+        images = set()
+        if target:
+            images.add(target)
+        else:
+            # gather all distinct images
+            ret, out, err = self.mon_command({
+                'prefix': 'config dump',
+                'format': 'json',
+            })
+            self.log.debug('config %s' % out)
+            config = json.loads(out)
+            for opt in config:
+                if opt['name'] == 'container_image':
+                    images.add(opt['value'])
+        if not len(images):
+            # default...
+            ret, out, err = self.mon_command({
+                'prefix': 'config get',
+                'who': 'mon.foo',
+                'key': 'container_image',
+            })
+            images.add(out.strip())
+        self.log.debug('Images: %s' % list(images))
+
+        image_versions = {}  # container image -> ceph version
+        for image in images:
+            image_versions[image] = self._get_container_image_id(image)
```

Comment:
This looks wasted. Why pull random images on random hosts? Can't we just store the image IDs persistently, or make the right hosts look for the image ids?

Reply:
The point here is that the image might change (e.g., ceph/ceph:v14.2 might get updated to the next point release).  In that case we need to query the registry every time.  We could use something like https://pypi.org/project/docker-registry-client/ to do this explicitly.. pulling to a random host was just a most expedient way forward.

I wonder, though: using a tag like ``v14.2`` is really a bad idea since you don't know what you're getting and will, over time, organically end up with a mixed set of versions deployed.  Perhaps we should just force the mode where you specify the ceph version you want and that's that.

Reply:
In a way, `14.2` is as bad as `:latest` (except that we're save against major version upgrades)

Kubernetes explicitly recommends [against](https://kubernetes.io/docs/concepts/configuration/overview/#container-images) `:latest`:

> Note: You should avoid using the :latest tag when deploying containers in production as it is harder to track which version of the image is running and more difficult to roll back properly.

Which is exactly what we're running into right now.

This is what's done by rook by default: 

https://github.com/rook/rook/blob/c5e79d4876a6acc27ea1a2a3fef27e7bd8a88145/cluster/examples/kubernetes/ceph/cluster.yaml#L23